### PR TITLE
Update kotlinx-metadata APIs for changes

### PR DESCRIPTION
**Pull request by ZacSweers**
No changelog yet: https://github.com/JetBrains/kotlin/blob/master/libraries/kotlinx-metadata/jvm/ChangeLog.md

### CODE REVIEW

**Review by Egorand — APPROVED** [review]
Wanna add a note to the "Unreleased" section of the changelog about the API changes?

**Comment by JakeWharton** [comment]
Yeah build that habit, Egor!

**Comment by Egorand** [comment]
@ZacSweers this would need to be rebased now that we've merged the 0.9.0 version bump: #1830.

File: interop/kotlinx-metadata/api/kotlinx-metadata.api [hunk 1]
```diff
@@ -1,19 +1,19 @@
...
-	public static final fun toKmClass (Ljava/lang/Class;)Lkotlinx/metadata/KmClass;
-	public static final fun toKmClass (Ljavax/lang/model/element/TypeElement;)Lkotlinx/metadata/KmClass;
-	public static final fun toKmClass (Lkotlin/Metadata;)Lkotlinx/metadata/KmClass;
-	public static final fun toKmClass (Lkotlin/reflect/KClass;)Lkotlinx/metadata/KmClass;
+	public static final fun readKotlinClassMetadata (Lkotlin/Metadata;Z)Lkotlinx/metadata/jvm/KotlinClassMetadata;
+	public static final fun toKmClass (Ljava/lang/Class;Z)Lkotlinx/metadata/KmClass;
+	public static final fun toKmClass (Ljavax/lang/model/element/TypeElement;Z)Lkotlinx/metadata/KmClass;
+	public static final fun toKmClass (Lkotlin/Metadata;Z)Lkotlinx/metadata/KmClass;
+	public static final fun toKmClass (Lkotlin/reflect/KClass;Z)Lkotlinx/metadata/KmClass;
 }
```
Comment (JakeWharton) [inline interop/kotlinx-metadata/api/kotlinx-metadata.api:7]: None of these functions have KotlinPoet types as parameter types or return types. Do they need to be public?

File: interop/kotlinx-metadata/src/main/kotlin/com/squareup/kotlinpoet/metadata/KotlinPoetMetadata.kt [hunk 2]
```diff
@@ -82,14 +99,16 @@ public inline fun <reified T : KotlinClassMetadata> Metadata.toKotlinClassMetada
...
 @KotlinPoetMetadataPreview
-public fun Metadata.readKotlinClassMetadata(): KotlinClassMetadata {
-  val metadata = KotlinClassMetadata.readStrict(asClassHeader())
-  checkNotNull(metadata) {
-    "Could not parse metadata! Try bumping kotlinpoet and/or kotlinx-metadata version."
+public fun Metadata.readKotlinClassMetadata(lenient: Boolean): KotlinClassMetadata {
+  return if (lenient) {
+    KotlinClassMetadata.readLenient(asClassHeader())
+  } else {
+    KotlinClassMetadata.readStrict(asClassHeader())
```
Comment (JakeWharton) [inline interop/kotlinx-metadata/src/main/kotlin/com/squareup/kotlinpoet/metadata/KotlinPoetMetadata.kt:110]: Does `asClassHeader()` do anything? It looks like it's copying from `Metadata` to `Metadata` (an identity transform?).

**Review by JakeWharton — APPROVED** [review]

File: interop/kotlinx-metadata/api/kotlinx-metadata.api (hunk 1 above)
Comment (ZacSweers) [reply interop/kotlinx-metadata/api/kotlinx-metadata.api:7]: We've historically offered a bit of a superset of APIs here with some helpers to make KM usage easier. It's become less necessary now though that the API on their side has caught up. Let me take a pass at cleaning these up in a followup PR?

File: interop/kotlinx-metadata/src/main/kotlin/com/squareup/kotlinpoet/metadata/KotlinPoetMetadata.kt (hunk 2 above)
Comment (ZacSweers) [reply interop/kotlinx-metadata/src/main/kotlin/com/squareup/kotlinpoet/metadata/KotlinPoetMetadata.kt:110]: Yep leftover from days past and probably the 0.9.0 merge with this branch. Let me clean up

File: interop/kotlinx-metadata/api/kotlinx-metadata.api (hunk 1 above)
Comment (JakeWharton) [reply interop/kotlinx-metadata/api/kotlinx-metadata.api:7]: Sounds good

File: interop/kotlinx-metadata/src/main/kotlin/com/squareup/kotlinpoet/metadata/KotlinPoetMetadata.kt (hunk 2 above)
Comment (ZacSweers) [reply interop/kotlinx-metadata/src/main/kotlin/com/squareup/kotlinpoet/metadata/KotlinPoetMetadata.kt]: [`c7a7df8` (#1766)](https://github.com/square/kotlinpoet/pull/1766/commits/c7a7df86caa614ea1899e00a0ff0718796967a47)

**Pull request merged**